components/compositing/touch.rs

```diff
@@ -198,56 +181,37 @@ impl TouchHandler {
-                debug_assert!((point.y as i64) < (i32::MAX as i64));
-                let cursor = DeviceIntPoint::new(point.x as i32, point.y as i32);
-                // Multiplying the initial velocity gives the fling a much more snappy feel
-                // and serves well as a poor-mans acceleration algorithm.
-                let velocity = (velocity * 2.0).with_max_length(FLING_MAX_SCREEN_PX);
-                self.state = Flinging { velocity, cursor };
-                TouchAction::NoAction
-            },
-            Panning { .. } => {
-                self.state = Nothing;
-                TouchAction::NoAction
-            },
-            Pinching => {
-                self.state = Panning {
-                    velocity: Vector2D::new(0.0, 0.0),
-                };
-                TouchAction::NoAction
-            },
-            Flinging { .. } => {
-                unreachable!("On touchup received, but already flinging.")
-            },
-            WaitingForScript => {
-                if self.active_touch_points.is_empty() {
+        match self.touch_count() {
+            0 => {
+                if let Panning { velocity } = self.state {
+                    if velocity.length().abs() >= FLING_MIN_SCREEN_PX {
+                        // TODO: point != old. Not sure which one is better to take as cursor for flinging.
+                        debug!(
+                            "Transitioning to Fling. Cursor is {point:?}. Old cursor was {old:?}. \
+                            Raw velocity is {velocity:?}."
+                        );
+                        debug_assert!((point.x as i64) < (i32::MAX as i64));
+                        debug_assert!((point.y as i64) < (i32::MAX as i64));
+                        let cursor = DeviceIntPoint::new(point.x as i32, point.y as i32);
+                        // Multiplying the initial velocity gives the fling a much more snappy feel
+                        // and serves well as a poor-mans acceleration algorithm.
+                        let velocity = (velocity * 2.0).with_max_length(FLING_MAX_SCREEN_PX);
+                        TouchAction::Flinging(velocity, cursor)
+                    } else {
+                        self.state = Nothing;
+                        TouchAction::NoAction
+                    }
+                } else {
                     self.state = Nothing;
-                    return TouchAction::Click;
+                    if !self.prevent_click {
+                        TouchAction::Click(point)
+                    } else {
+                        TouchAction::NoAction
+                    }
                 }
-                TouchAction::NoAction
             },
-            DefaultPrevented | MultiTouch => {
-                if self.active_touch_points.is_empty() {
-                    self.state = Nothing;
-                }
+            _ => {
+                self.state = Touching;
```

Comment:
- if we have 2 (maybe 3?) or more fingers here, then we should stay in MultiTouch I think. I guess we don't have any Multitouch gestures right now, it doesn't matter too much.
- if we have 1 finger, then we could be coming from pinch or MultiTouch. I guess transitioning to Touching is correct. We must ensure that no `Click` event will be triggered when the final finger is lifted, but this is already handled by setting `prevent_click = true` in on_touch_down (so nothing to do for this point).